ssl/ssl_lib.c

```diff
@@ -3369,18 +3369,21 @@ void ssl_update_cache(SSL *s, int mode)
         && (!s->hit || SSL_IS_TLS13(s))) {
         /*
          * Add the session to the internal cache. In server side TLSv1.3 we
-         * normally don't do this because its a full stateless ticket with only
-         * a dummy session id so there is no reason to cache it, unless:
+         * normally don't do this because by default its a full stateless ticket
+         * with only a dummy session id so there is no reason to cache it,
+         * unless:
          * - we are doing early_data, in which case we cache so that we can
          *   detect replays
          * - the application has set a remove_session_cb so needs to know about
          *   session timeout events
+         * - SSL_OP_NO_TICKET is set in which case it is a stateful ticket
          */
         if ((i & SSL_SESS_CACHE_NO_INTERNAL_STORE) == 0
                 && (!SSL_IS_TLS13(s)
                     || !s->server
                     || s->max_early_data > 0
-                    || s->session_ctx->remove_session_cb != NULL))
+                    || s->session_ctx->remove_session_cb != NULL
+                    || (s->options & SSL_OP_NO_TICKET) != 0))
```

Comment:
With the `!s->server` test above (unchanged in this PR), it looks like we're adding all sessions to the internal store on the client.  But I'm not aware of any way to use these... Why do we do that?

Reply:
Client side caching is always off by default. We only cache on the client if the application has explicitly requested it. It *is* possible to use the client side cache (and I have seen questions in the past on openssl-users from people using it). It's not exactly an easy or efficient interface though. You can call the `SSL_CTX_sessions()` function to get hold of the cache. The cache is stored as an LHASH keyed by session_id. That's not particularly useful on the client, but you can (for example) still iterate over all sessions in the cache using `lh_SSL_SESSION_doall()` which could (maybe) be used to search for a suitable session for resumption.

Mainly this is retained for backwards compatibility.